README.md
# DiagnosticTool
This tool is intended to help automate Macintosh diagnostics. It relies heavily on system_profiler to pull relevent information about the host machine and generates a report with this information. See the sampleOutput.txt for a better understanding of what it does.

# Caveats
Depending on the macOS version, errors may be encountered with some of the functions of this tool. I've tried my best to mitigate and/or work around some of them, and this is a work in progress. If you have any feedback please drop me a line.

# Version Compatability
Tested extensively on macOS Catalina, and most outputs are generally as expected across all Mac models. However it is quite challenging to outline every caveat in every specific machine across every single software version -- so expect some bugs here and there. Again, feel free to report any bugs you may encounter, and also any ideas for improvements, etc.

# Known Issues
macOS Big Sur
- some of the outputs print blank spaces in the final report, I haven't been able to test as extensively on 11.X yet

Legacy versions of macOS/OS X
- the information system_profiler provides varies between software versions, so not every output will work on every software version
- I've gone to great length to ensure that most of the important information is printed across various machines, but most of my focus has been on later/current versions of macOS


# Thanks to

krypted for swiftwarrantylookup, find it here: https://github.com/krypted/swiftwarrantylookup

Binary Fruit for DriveDx, find it here: https://binaryfruit.com/drivedx

Sqwarq for DetectX Swift, find it here: https://sqwarq.com/detectx/


changelog.md
# Change-log:

# v2.2.1 - 11/15/21
additions
- added options (to be expanded in future versions)
- first release


# v2.2 - 6/15/21
bug fixes - big sur
- fixed graphics info section regarding metal support not printing correctly
- fixed battery info section to correctly detail the information available on big sur

changes
- reintroduced output of system profile in xml/json
- now exports relevent facets to a folder named after the host machine's serial number
    - this can be annoying if this extra output is not desiered
    - comment out "OUTPUTSTART" near the end of MAIN if this additional output is undesired



# v2.2 - In progress - 6/11/21

bug fixes - big sur
- graphics info section reporting "Metal: Not Supported" incorrectly
    - system_profiler does not include this section since it will only run on metal-supported graphics natively
    - introduced nested if statements to evaulate software version if that section is not found, needs further testing
- battery info section
    - added conditionals to fix blank prints, tested on 16-inch mbp

changes
- scaled down shutdown log to parse only 10 days instead of 14
    - still trying to find a good balance between collecting enough data and how long it takes to return data
- added boot mode printing in the system overview section
    - seems to be working but needs further testing for version compatibility
- removed version numbers from the notifications
- cleaned up the osascript prompts and their buttons



# v2.1 beta2 —update 5/24/21
improved functionality of network information printing
- conditional logic dependent on active network interface
    - support on Mac Pro needs further testing
        - there’s still some issues with network configuration info
        - it may be too variable between different machines to depend on it working 100% of the time
- fixed display type blank printing
    - will determine if display type is a TV, or print nothing
- fixed bluetooth version printing
- added the product description back from the original diag script
    - issues with this
        - on later versions of macOS the program may not run
        - to compensate, there’s a fallback that it will use the system profiler name description
- tested the script on an M1 MBA running 11.2


# v2.1 beta2 — 5/18/21
Version compatibility issues resolved
- Fixed some more blank prints on legacy versions
- Better functionality in network info printing
    - Still needs a bit of work, its not very dynamic
        - Think I broke IPv4 info
        - Need to make this conditional



# v2.1 beta — 5/17/21
Squashed some bugs
- blank spaces (still more to go)
- Double prints (probably more I haven’t thought of yet)
- Shutdown log not running sometimes
- Malware scan not running sometimes

Added support for
- Json output in macOS Catalina
    - Need to add logic for Big Sur and test it
- Xml output in legacy versions of macOS
    - Needs testing

Future versions
- Going to need to include separate functions for legacy compatibility 
- Will include logic that does something with the outFiles (json output)
- Will include a python module to parse xml output and convert it to csv
- Progress bar cleanup
    - Needs more steps so it looks more realistic
    - Adding description of progress for the bored user to read
- Background heavier functions to make better use of parallel processing
    - User directory usage (du is kind of heavy)
    - I imagine disk usage will be more expensive on rotational mediums



# v2.1 alpha — 5/10/21
Complete overhaul of the code
- Dependent functions are now properly functional-ized
    - Use local variables and global where appropriate
- Error output is suppressed from within rather than from without
- Functions to evaluate proper course for the program to run
    - These depend on multiple evaluations at different points to determine which path to completion should be followed
- Functions fork off of main where appropriate
- Wider compatibility achieved through variable functions for different machines
- Added notifications for the end user as the program progresses

Support now added for:
- Upgradeable memory
- Metal-capable graphics
- Proper display type & resolution printing
- Properly handled dual graphics MacBooks
- Properly handled dual graphics in Mac Pro 6,1
- AC Charger information
- Properly handled IPv6 configurations
- Filesystem info, ownership, device name, dev node, and attachment protocol
- FileVault status printing

Running on:
- MacBooks (13 inch & smaller models)
- MacBooks (15/16 inch & dual graphics models)
- iMac
- Mac Pro 5,1 (and possibly earlier)
- Mac Pro 6,1 (dual graphics models included)

Need to re-introduce external functions now:
- SMART stats??
    - Going to run these in the background
        - These will output to /tmp so the end user will never see them
        - These will run in parallel so that the program will be more efficient
        - The program will not exit before all of the external functions have been completed
        - The external functions will concatenate to the final report upon completion of main


# v2.0.4b — 5/6/21
Working on adding desktop functionality
- Conditional functions for desktop-related functions
- Fixed conditional for laptops with discreet graphics




legacy versions/diagv2.1alpha.sh
#! /bin/bash

# Diagnostic Tool v2.1 alpha

# keep an eye on outputs
# may have to work with the variable naming scheme
# something about local vs global variables

# also maybe rename the $path variable that points to /tmp

# have to write additional logic for MacPro6,1 w/ dual gfx

# contains instructions for LAPTOPS
function LAPTOPMEMORYINFO(){ 


	local memBanks=$(
		system_profiler SPMemoryDataType \
		| grep -A 7 "BANK" \
		| awk '{print $1,$2,$3,$4}'
		)

	echo "$memBanks" >> ~/Desktop/$serial.txt

	local ramECC=$(
		system_profiler SPMemoryDataType \
		| grep "ECC" \
		| awk '{print $2}'
		)

	echo "--" >> ~/Desktop/$serial.txt # writes proper spacing to report
	
	echo "ECC: $ramECC" >> ~/Desktop/$serial.txt

	local upgradeRAM=$(
		system_profiler SPMemoryDataType \
		| grep "Upgradeable Memory:" \
		| awk '{print $3}'
		)

	echo "Upgradeable Memory: $upgradeRAM" >> ~/Desktop/$serial.txt

	local ramSize=$(
		system_profiler SPHardwareDataType \
		| grep "Memory" \
		| awk '{print $2,$3}'
		)

	echo "Total RAM Installed: $ramSize" >> ~/Desktop/$serial.txt

	local swapUsed=$(
		sysctl vm.swapusage \
		| awk '{print $2,$3,$4,$5,$6,$7,$8,$9,$10,$11}'
		)

	echo "Swap Usage: $swapUsed" >> ~/Desktop/$serial.txt


}

# contains instructions for DESKTOPS
function DESKTOPMEMORYINFO(){

	local memDimms=$(
		system_profiler SPMemoryDataType \
		| grep -A 7 "DIMM" \
		| awk '{print $1,$2,$3,$4}'
		)

	echo "$memDimms" >> ~/Desktop/$serial.txt

	local ramECC=$(
		system_profiler SPMemoryDataType \
		| grep "ECC" \
		| awk '{print $2}'
		)

	echo "--" >> ~/Desktop/$serial.txt # writes proper spacing to report
	 
	echo "ECC: $ramECC" >> ~/Desktop/$serial.txt

	local upgradeRAM=$(
		system_profiler SPMemoryDataType \
		| grep "Upgradeable Memory:" \
		| awk '{print $3}'
		)

	echo "Upgradeable Memory: $upgradeRAM" >> ~/Desktop/$serial.txt

	local ramSize=$(
		system_profiler SPHardwareDataType \
		| grep "Memory" \
		| awk '{print $2,$3}'
		)

	echo "Total RAM Installed: $ramSize" >> ~/Desktop/$serial.txt

	local swapUsed=$(
		sysctl vm.swapusage \
		| awk '{print $2,$3,$4,$5,$6,$7,$8,$9,$10,$11}'
		)

	echo "Swap Usage: $swapUsed" >> ~/Desktop/$serial.txt

}

# contains conditional to determine if host is laptop or desktop
function MEMORYINFOSTART {

	if [ "$modelType" == "MacBook" ]; then LAPTOPMEMORYINFO
	else DESKTOPMEMORYINFO
	fi

}

# contains instructions for DESKTOPS
function DESKTOPGRAPHICSINFO(){

	local gpu=$(
		system_profiler SPDisplaysDataType \
		| grep "Chipset Model:" \
		| awk '{print $3,$4,$5}'
		)

	echo "Graphics: $gpu" >> ~/Desktop/$serial.txt

	local bus=$(
		system_profiler SPDisplaysDataType \
		| grep "Bus:" \
		| awk '{print $2}'
		)

	echo "Connection: $bus" >> ~/Desktop/$serial.txt

	local slot=$(
		system_profiler SPDisplaysDataType \
		| grep "Slot:" \
		| awk '{print $2}'
		)
	
	# slot info not available on every desktop mac model
	# this if statement tests to see if the variable $slot is available
	# if so, it will print that info to report, if not it will do nothing
	# this way it does not leave a blank in the final report

	if test -z "$slot"; then echo "slot information not available for this device" > /dev/null
	else echo "Slot: $slot" >> ~/Desktop/$serial.txt
	fi


	local laneWidth=$(
		system_profiler SPDisplaysDataType \
		| grep "PCIe Lane Width:" \
		| awk '{print $4}'
		)

	echo "PCIe Lane Width: $laneWidth" >> ~/Desktop/$serial.txt

	local gpuVRAM=$(
		system_profiler SPDisplaysDataType \
		| grep "VRAM" \
		| awk '{print $1,$2,$3,$4,$5,$6}'
		)

	echo "$gpuVRAM" >> ~/Desktop/$serial.txt

	local vendor=$(
		system_profiler SPDisplaysDataType \
		| grep "Vendor:" \
		| awk '{print $2}'
		)

	echo "Manufacturer: $vendor" >> ~/Desktop/$serial.txt

	local metal=$(
		system_profiler SPDisplaysDataType \
		| grep "Metal:" \
		| awk '{print $2}'
		)

	if [ "$metal" == "Supported," ]
		then echo "Metal: Supported" >> ~/Desktop/$serial.txt
		else echo "Metal: Not Supported" >> ~/Desktop/$serial.txt
	fi

	# displays data
	local dispType=$(
		system_profiler SPDisplaysDataType \
		| grep "Display Type:" \
		| awk '{print $3}'
		)

	echo "Display Type: $dispType" >> ~/Desktop/$serial.txt

	echo "Display Resolution: $dispRes" >> ~/Desktop/$serial.txt

	local framebuffer=$(
		system_profiler SPDisplaysDataType \
		| grep "Framebuffer Depth:" \
		| awk '{print $3,$4,$5,$6,$7}'
		)

	echo "Framebuffer Depth: $framebuffer" >> ~/Desktop/$serial.txt

}

# contains instructions for LAPTOPS with integrated graphics ONLY
function LAPTOPINTEGRATEDGRAPHICSINFO(){

	local intGPU=$(
		system_profiler SPDisplaysDataType \
		| grep -B 3 "Bus: Built-In" \
		| grep "Chipset Model" \
		| awk '{print $3,$4,$5,$6}'
		)

	echo "Integrated Graphics: $intGPU" >> ~/Desktop/$serial.txt

	local vramInt=$(
		system_profiler SPDisplaysDataType \
		| grep "VRAM (Dynamic, Max)" \
		| awk '{print $4,$5}'
		)

	echo "VRAM: $vramInt" >> ~/Desktop/$serial.txt

	local metal=$(
		system_profiler SPDisplaysDataType \
		| grep "Metal:" \
		| awk '{print $2}'
		)
	
	if [ "$metal" == "Supported," ]
		then echo "Metal: Supported" >> ~/Desktop/$serial.txt
		else echo "Metal: Not Supported" >> ~/Desktop/$serial.txt
	fi

	
	# displays info
	local dispType=$(
		system_profiler SPDisplaysDataType \
		| grep "Display Type" \
		| awk '{print $3,$4,$5}'
		)

	echo "Display Type: $dispType" >> ~/Desktop/$serial.txt

	local dispRes=$(
		system_profiler SPDisplaysDataType \
		| grep "Resolution" \
		| awk '{print $2,$3,$4,$5}'
		)

	echo "Resolution: $dispRes" >> ~/Desktop/$serial.txt

	local framebuffer=$(
		system_profiler SPDisplaysDataType \
		| grep "Framebuffer Depth:" \
		| awk '{print $3,$4,$5,$6,$7}'
		)

	echo "Framebuffer Depth: $framebuffer" >> ~/Desktop/$serial.txt

}

# contains instructions for LAPTOPS with integrated AND discreet graphics
function LAPTOPDUALGRAPHICSINFO(){

	local intGPU=$(
		system_profiler SPDisplaysDataType \
		| grep -B 3 "Built-In" | grep "Chipset Model" \
		| awk '{print $3,$4,$5,$6}'
		)

	local vramInt=$(
		system_profiler SPDisplaysDataType \
		| grep "VRAM (Dynamic, Max)" \
		| awk '{print $4,$5}'
		)

	echo "Integrated Graphics: $intGPU $vramInt" >> ~/Desktop/$serial.txt

	local dGPU=$(
		system_profiler SPDisplaysDataType \
		| grep -B 3 "Bus: PCIe" \
		| grep "Chipset Model:" \
		| awk '{print $3,$4,$5,$6}'
		)

	local vramD=$(
		system_profiler SPDisplaysDataType \
		| grep "VRAM (Total):" \
		| awk '{print $3,$4}'
		)

	echo "Discreet Graphics: $dGPU $vramD" >> ~/Desktop/$serial.txt
	
	local metal=$(
		system_profiler SPDisplaysDataType \
		| grep "Metal:" \
		| awk '{print $2}'
		)
	
	if [ "$metal" == "Supported," ]
		then echo "Metal: Supported" >> ~/Desktop/$serial.txt
		else echo "Metal: Not Supported" >> ~/Desktop/$serial.txt
	fi
		

	# displays info
	local dispType=$(
		system_profiler SPDisplaysDataType \
		| grep "Display Type" \
		| awk '{print $3,$4,$5}'
		)

	echo "Display Type: $dispType" >> ~/Desktop/$serial.txt

	local dispRes=$(
		system_profiler SPDisplaysDataType \
		| grep "Resolution" \
		| awk '{print $2,$3,$4,$5}'
		)

	echo "Resolution: $dispRes" >> ~/Desktop/$serial.txt

	local framebuffer=$(
		system_profiler SPDisplaysDataType \
		| grep "Framebuffer Depth:" \
		| awk '{print $3,$4,$5,$6,$7}'
		)

	echo "Framebuffer Depth: $framebuffer" >> ~/Desktop/$serial.txt

}

# contains the test conditions that determine whether a laptop has dual graphics or not
function GRAPHICSTYPETEST {

	gfxSwitching=$(
		system_profiler SPDisplaysDataType \
		| grep "Automatic Graphics Switching:"
		)

	if test -z "$gfxSwitching" 
		then
		  LAPTOPINTEGRATEDGRAPHICSINFO
		else
		  LAPTOPDUALGRAPHICSINFO 
	fi


}

# contains conditional to determine which graphics function to execute
function GRAPHICSINFOSTART {

	if [ "$modelType" == "MacBook" ]; then GRAPHICSTYPETEST
	else DESKTOPGRAPHICSINFO
	fi

}

# contains instructions for printing battery information in LAPTOPS
function BATTERYINFO(){

	echo "Battery Installed: Yes" >> ~/Desktop/$serial.txt

	local batteryCondition=$(
		system_profiler SPPowerDataType \
		| grep "Condition" \
		| awk '{print $2,$3}'
		)

	echo "Battery Condition: $batteryCondition" >> ~/Desktop/$serial.txt

	local batteryCycles=$(
		system_profiler SPPowerDataType \
		| grep "Cycle Count" \
		| awk '{print $3}'
		)

	echo "Battery Cycle Count: $batteryCycles" >> ~/Desktop/$serial.txt

	local batteryCapacity=$(
		system_profiler SPPowerDataType \
		| grep "Full Charge Capacity (mAh)" \
		| awk '{print $5}'
		)

	echo "Full Charge Capacity: $batteryCapacity" >> ~/Desktop/$serial.txt

	local batteryCharge=$(
		system_profiler SPPowerDataType \
		| grep "Charge Remaining (mAh)" \
		| awk '{print $4}'
		)

	echo "Charge Remaining: $batteryCharge" >> ~/Desktop/$serial.txt

	local batteryAmp=$(
		system_profiler SPPowerDataType \
		| grep "Amperage (mA):" \
		| awk '{print $3}'
		)

	echo "Amperage (mA): $batteryAmp" >> ~/Desktop/$serial.txt

	local batteryVolt=$(
		system_profiler SPPowerDataType \
		| grep "Voltage (mV)" \
		| awk '{print $3}'
		)

	echo "Voltage (mV): $batteryVolt" >> ~/Desktop/$serial.txt
	
	chargerConnected=$(
		system_profiler SPPowerDataType \
		| grep -A 2 "AC Charger Information" \
		| grep "Connected:" \
		| awk '{print $2}'
		)
		
	if [ "$chargerConnected" == "Yes" ]
		then
			echo "AC Charger Connected: $chargerConnected" >> ~/Desktop/$serial.txt;		
			ACCHARGERINFO
	else
		echo "AC Charger Connected: $chargerConnected" >> ~/Desktop/$serial.txt
	fi

}

# contains instructions for printing AC Charger information in LAPTOPS, only if connected, as defined in BATTERYINFO
function ACCHARGERINFO(){

	local charging=$(
		system_profiler SPPowerDataType \
		| grep -A 11 "AC Charger Information" \
		| grep "Charging:" \
		| awk '{print $2}'
		)

	echo "Charging: $charging" >> ~/Desktop/$serial.txt

	local chargingWattage=$(
		system_profiler SPPowerDataType \
		| grep -A 11 "AC Charger Information" \
		| grep "Wattage (W):" \
		| awk '{print $3}'
		)
	
	echo "Wattage (W): $chargingWattage" >> ~/Desktop/$serial.txt

	local chargerName=$(
		system_profiler SPPowerDataType \
		| grep -A 11 "AC Charger Information" \
		| grep "Name:" \
		| awk '{print $2,$3,$4,$5,$6}'
		)

	echo "Name: $chargerName" >> ~/Desktop/$serial.txt

	local chargerVendor=$(
		system_profiler SPPowerDataType \
		| grep -A 11 "AC Charger Information" \
		| grep "Manufacturer:" \
		| awk '{print $2,$3,$4,$5,$6}'
		)

	echo "Manufacturer: $chargerVendor" >> ~/Desktop/$serial.txt

}

# contains instructions for printing AC Power information in DESKTOPS
function ACPOWERINFO(){

	echo "Power Source: AC Power" >> ~/Desktop/$serial.txt

	local restartOnLoss=$(
		system_profiler SPPowerDataType \
		| grep "Automatic Restart on Power Loss:" \
		| awk '{print $1,$2,$3,$4,$5,$6}'
		)

	echo "$restartOnLoss" >> ~/Desktop/$serial.txt

	local wakeOnLAN=$(
		system_profiler SPPowerDataType \
		| grep "Wake on LAN:" \
		| awk '{print $4}'
		)

	echo "Wake on LAN: $wakeOnLAN" >> ~/Destkop/$serial.txt

	local gpuSwitch=$(
		system_profiler SPPowerDataType \
		| grep -A 18 "AC Power" \
		| grep "GPUSwitch" \
		| awk '{print $2}'
		)

	echo "GPU Switch: $gpuSwitch" >> ~/Desktop/$serial.txt
	
	local sysSleepTimer=$(
		system_profiler SPPowerDataType \
		| grep "System Sleep Timer" \
		| awk '{print $5}'
		)
	
	echo "System Sleep Timer (Minutes): $sysSleepTimer" >> ~/Desktop/$serial.txt
	
	local diskSleepTimer=$(
		system_profiler SPPowerDataType \
		| grep "Disk Sleep Timer" \
		| awk '{print $5}'
		)
		
	echo "Disk Sleep Timer (Minutes): $diskSleepTimer" >> ~/Desktop/$serial.txt

	local ups=$(
		system_profiler SPPowerDataType \
		| grep "UPS Installed:" \
		| awk '{print $3}'
		)

	echo "UPS Installed: $ups" >> ~/Desktop/$serial.txt

}

# contains instructions on which power information function to execute
function POWERINFOSTART {

	if [ "$modelType" == "MacBook" ]; then BATTERYINFO
	else ACPOWERINFO
	fi

}

# contains instructions for printing wifi signal to report
function WIRELESSSIGNALINFO {

	/System/Library/PrivateFrameworks/Apple80211.framework/Versions/Current/Resources/airport -I >> ~/Desktop/$serial.txt

}

# contains instructions for printing disk usage to report
function DISKUSAGE(){

	df -H /System/Volumes/Data \
	| awk '{printf"%-25s%-10s%-10s%-10s\n", $9,$3,$4,$5}' \
	>> ~/Desktop/$serial.txt

}

# contains instructions for printing home folder usage to report
# this should be an external (background operation) in the next version
function USERDIRECTORYUSAGE(){

	cd ~

	du -hsc * >> ~/Desktop/$serial.txt

}




function MAIN {

	# grab serial number and create report named $serial.txt
	serial=$(
		system_profiler SPHardwareDataType \
		| grep "Serial Number (system)" \
		| awk '{print $4}'
		)

	echo "Diagnostic Tool v2.1 alpha -- The Mac Support Store" > ~/Desktop/$serial.txt

	echo >> ~/Desktop/$serial.txt

	echo "----------------" >> ~/Desktop/$serial.txt




	# system overview
	echo "System Profile:" >> ~/Desktop/$serial.txt

	modelName=$(
		system_profiler SPHardwareDataType \
		| grep "Model Name" \
		| awk '{print $3,$4,$5,$6}'
		)

	echo "Mac Model: $modelName" >> ~/Desktop/$serial.txt

	modelID=$(
		system_profiler SPHardwareDataType \
		| grep "Model Identifier" \
		| awk '{print $3}'
		)

	echo "Model Identifier: $modelID" >> ~/Desktop/$serial.txt

	echo "Serial Number: $serial" >> ~/Desktop/$serial.txt




	# macOS version and build number
	softwareVersion=$(
		system_profiler SPSoftwareDataType \
		| grep "System Version" \
		| awk '{print $3,$4,$5}'
		)

	echo "Software Version: $softwareVersion" >> ~/Desktop/$serial.txt

	# boot rom version
	bootRom=$(
		system_profiler SPHardwareDataType \
		| grep "Boot ROM" \
		| awk '{print $4,$5,$6,$7}'
		)

	echo "Boot ROM Version: $bootRom" >> ~/Desktop/$serial.txt

	# system integrity protection status
	csrutilStatus=$(
		csrutil status
		)

	echo "$csrutilStatus" >> ~/Desktop/$serial.txt




	# chipset info
	echo "----------------" >> ~/Desktop/$serial.txt

	echo "Chipset:" >> ~/Desktop/$serial.txt

	procName=$(
		sysctl machdep.cpu.brand_string \
		| awk '{print $2,$3,$4}'
		)

	echo "Processor Type: $procName" >> ~/Desktop/$serial.txt

	procSpeed=$(
		system_profiler SPHardwareDataType \
		| grep "Processor Speed" \
		| awk '{print $3,$4,$5,$6,$7}'
		)

	echo "Processor Speed: $procSpeed" >> ~/Desktop/$serial.txt

	procNumber=$(
		system_profiler SPHardwareDataType \
		| grep "Number of Processors" \
		| awk '{print $4}'
		)

	echo "Number of Processors: $procNumber" >> ~/Desktop/$serial.txt

	procCores=$(
		system_profiler SPHardwareDataType \
		| grep "Total Number of Cores" \
		| awk '{print $5}'
		)

	echo "Number of Processing Cores: $procCores" >> ~/Desktop/$serial.txt

	procThread=$(
		system_profiler SPHardwareDataType \
		| grep "Hyper-Threading" \
		| awk '{print $3}'
		)

	echo "Hyper-Threading Technology: $procThread" >> ~/Desktop/$serial.txt




	# memory info section
	modelType=$(
		system_profiler SPHardwareDataType \
		| grep "Model Name:" \
		| awk '{print $3}'
		)

	echo "----------------" >> ~/Desktop/$serial.txt

	echo "Memory Info:
	--" >> ~/Desktop/$serial.txt

	MEMORYINFOSTART




	# graphics info section
	echo "----------------" >> ~/Desktop/$serial.txt

	echo "Graphics & Displays:" >> ~/Desktop/$serial.txt

	GRAPHICSINFOSTART




	# power info section
	echo "----------------" >> ~/Desktop/$serial.txt

	echo "Power Info:" >> ~/Desktop/$serial.txt

	POWERINFOSTART




	# wireless signal sample
	echo "----------------" >> ~/Desktop/$serial.txt

	echo "Wireless Statistics:" >> ~/Desktop/$serial.txt

	WIRELESSSIGNALINFO




	# wireless interface info section
	echo "----------------" >> ~/Desktop/$serial.txt

	echo "Network Info:" >> ~/Desktop/$serial.txt

	intName=$(
		system_profiler SPNetworkDataType \
		| head -25 \
		| grep "ConfirmedInterfaceName" \
		| awk '{print $2}'
		)

	echo "Wireless Interface: $intName" >> ~/Desktop/$serial.txt

	intType=$(
		system_profiler SPNetworkLocationDataType \
		| head -15 \
		| grep "Type" \
		| awk '{print $2}'
		)

	echo "Interface Type: $intType" >> ~/Desktop/$serial.txt

	wifiMacAddr=$(
		system_profiler SPNetworkDataType \
		| head -20 \
		| grep "ARPResolvedHardwareAddress" \
		| awk '{print $2}'
		)

	echo "MAC Address: $wifiMacAddr" >> ~/Desktop/$serial.txt

	ipAddr=$(
		system_profiler SPNetworkDataType \
		| head -20 \
		| grep "IPv4 Addresses" \
		| awk '{print $3}'
		)

	echo "IPv4 Address: $ipAddr" >> ~/Desktop/$serial.txt

	ipv4ConfigType=$(
		system_profiler SPNetworkDataType \
		| head -30 \
		| grep -A 10 "IPv6" \
		| grep "Configuration Method" \
		| awk '{print $3}'
		)

	echo "IPv4 Configuration Type: $ipv4ConfigType" >> ~/Desktop/$serial.txt

	ipv6ConfigType=$(
		system_profiler SPNetworkDataType \
		| head -30 \
		| grep -A 1 "IPv6" \
		| grep "Configuration" \
		| awk '{print $3}'
		)

	echo "IPv6 Configuration Type: $ipv6ConfigType" >> ~/Desktop/$serial.txt




	# bluetooth info section
	echo "----------------" >> ~/Desktop/$serial.txt

	echo "Bluetooth:" >> ~/Desktop/$serial.txt

	btMan=$(
		system_profiler SPBluetoothDataType \
		| head -15 \
		| grep "Manufacturer" \
		| awk '{print $2}'
		)

	echo "Bluetooth Manufacturer: $btMan" >> ~/Desktop/$serial.txt

	btRev=$(
		system_profiler SPBluetoothDataType \
		| head -25 \
		| grep "Bluetooth Core Spec" \
		| awk '{print $4,$5,$6}'
		)

	echo "Bluetooth Revision: $btRev" >> ~/Desktop/$serial.txt

	btPower=$(
		system_profiler SPBluetoothDataType \
		| grep "Bluetooth Power" \
		| awk '{print $3}'
		)

	echo "Bluetooth Power: $btPower" >> ~/Desktop/$serial.txt

	btDiscover=$(
		system_profiler SPBluetoothDataType \
		| head -15 \
		| grep "Discoverable" \
		| awk '{print $2}'
		)

	echo "Discovery: $btDiscover" >> ~/Desktop/$serial.txt

	handoff=$(
		system_profiler SPBluetoothDataType \
		| head -15 \
		| grep "Handoff Supported" \
		| awk '{print $3}'
		)

	echo "Handoff Supported: $handoff" >> ~/Desktop/$serial.txt

	btHostname=$(
		system_profiler SPBluetoothDataType \
		| head -15 \
		| grep "Name" \
		| awk '{print $2,$3,$4,$5}'
		)

	echo "Hostname: $btHostname" >> ~/Desktop/$serial.txt

	btMacAddr=$(
		system_profiler SPBluetoothDataType \
		| head -15 \
		| grep "Address" \
		| awk '{print $2}'
		)

	echo "MAC Address: $btMacAddr" >> ~/Desktop/$serial.txt




	# disk info section
	echo "----------------" >> ~/Desktop/$serial.txt

	echo "Disk Overview:" >> ~/Desktop/$serial.txt

	diskType=$(
		system_profiler SPStorageDataType \
		| head -20 \
		| grep "Medium Type" \
		| awk '{print $3}'
		)

	echo "Storage Medium: $diskType" >> ~/Desktop/$serial.txt

	fsType=$(
		system_profiler SPStorageDataType \
		| head -20 \
		| grep "File System:" \
		| awk '{print $3,$4,$5,$6}'
		)

	echo "File System: $fsType" >> ~/Desktop/$serial.txt

	writable=$(
		system_profiler SPStorageDataType \
		| head -20 \
		| grep "Writable:" \
		| awk '{print $2}'
		)

	echo "Writable: $writable" >> ~/Desktop/$serial.txt

	ignoreOwnership=$(
		system_profiler SPStorageDataType \
		| head -10 \
		| grep "Ignore Ownership:" \
		| awk '{print $3}'
		)

	echo "Ignore Ownership: $ignoreOwnership" >> ~/Desktop/$serial.txt

	bsdName=$(
		system_profiler SPStorageDataType \
		| head -20 \
		| grep "BSD Name:" \
		| awk '{print $3}'
		)

	echo "BSD Name: $bsdName" >> ~/Desktop/$serial.txt

	devName=$(
		system_profiler SPStorageDataType \
		| head -20 \
		| grep "Device Name:" \
		| awk '{print $3,$4,$5,$6,$7,$8,$9}'
		)

	echo "Device Name: $devName" >> ~/Desktop/$serial.txt


	attachmentProtocol=$(
		system_profiler SPStorageDataType \
		| head -21 \
		| grep "Protocol:" \
		| awk '{print $2,$3}'
		)

	echo "Protocol: $attachmentProtocol" >> ~/Desktop/$serial.txt

	attachmentLocation=$(
		system_profiler SPStorageDataType \
		| head -21 \
		| grep "Internal:" \
		| awk '{print $2}'
		)

	echo "Internal: $attachmentLocation" >> ~/Desktop/$serial.txt

	DISKUSAGE




	# user directory usage
	echo "----------------" >> ~/Desktop/$serial.txt

	echo "Home Folder Usage:" >> ~/Desktop/$serial.txt

	USERDIRECTORYUSAGE


}

MAIN 2> /dev/null

